extensions/AWS/S3/S3Storage.cs

```diff
@@ -0,0 +1,346 @@
+        {
+            BucketName = this._bucketName,
+            Key = objName,
+            InputStream = streamContent
+        }, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+        this._log.LogTrace("Object {0} ready, size {1}", objName, len);
+    }
+
+    /// <inheritdoc />
+    public async Task<StreamableFileContent> ReadFileAsync(
+        string index,
+        string documentId,
+        string fileName,
+        bool logErrIfNotFound = true,
+        CancellationToken cancellationToken = default)
+    {
+        var directoryName = JoinPaths(index, documentId);
+        var objName = $"{directoryName}/{fileName}";
+
+        try
+        {
+            using (var response = await this._client.GetObjectAsync(
+                new GetObjectRequest
+                {
+                    BucketName = this._bucketName,
+                    Key = objName,
+                },
+                cancellationToken
+            ).ConfigureAwait(false))
+            {
+                var memoryStream = new MemoryStream();
+                var responseStream = response.ResponseStream;
+
+                await using (responseStream.ConfigureAwait(false))
+                {
+                    await responseStream.CopyToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+
+                    return new StreamableFileContent(
+                        fileName,
+                        response.ContentLength,
+                        response.Headers.ContentType,
+                        response.LastModified,
+                        async () =>
+                        {
+                            memoryStream.Seek(0, SeekOrigin.Begin);
+                            return await Task.FromResult(memoryStream).ConfigureAwait(false);
+                        }
+                    );
+                }
+            }
+        }
+        catch (AmazonS3Exception e) when (e.StatusCode == HttpStatusCode.NotFound)
+        {
+            if (logErrIfNotFound)
+            {
+                this._log.LogInformation("File not found: {0}", objName);
+            }
+
+            throw new DocumentStorageFileNotFoundException("File not found", e);
+        }
+    }
+
+    /// <summary>
+    /// Generates a pre-signed URL for accessing an object stored in S3 for a limited time.
+    /// </summary>
+    /// <param name="index">The index directory in which the object resides.</param>
+    /// <param name="documentId">The document directory under the index where the object resides.</param>
+    /// <param name="fileName">The name of the file/object for which to generate the URL.</param>
+    /// <param name="validDuration">The duration for which the URL should remain valid.</param>
+    /// <returns>A pre-signed URL for accessing the object.</returns>
+    public async Task<string> GeneratePreSignedURLAsync(string index, string documentId, string fileName, TimeSpan validDuration)
```

Comment:
Also this method is not part of IDocumentStorage interface, I need it in my project to be able to show links to the uploaded files as citations in the LLM answers. Do you have any advice how to stay in conformation with IDocumentStorage and still have this feature?

Reply:
As a rule of thumb I would avoid extra public methods in the repo, to reduce the cost of maintaining it - perhaps consider a static extension method. I'll take a look any way, there's always exceptions :-)

Reply:
✅ I removed this method. In my project I had to implement it separately with another instance of AmazonS3Client because if I create extension on this S3Storage it still needs to access private variables which is not allowed by compiler.